mbed_build/_internal/config/config_action.py

```diff
@@ -0,0 +1,79 @@
+#
+# Copyright (C) 2020 Arm Mbed. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+"""This module contains config actions, which will mutate configuration.
+
+Config actions are built from entries found in JSON files used by MbedOS applications.
+Some of those entries aren't simple overrides, but involve more complex operations
+like addition or removal of items from lists.
+"""
+from abc import ABC, abstractmethod, abstractclassmethod
+from dataclasses import dataclass
+from typing import Any, Optional
+
+# Fix circular dependency problem
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from mbed_build._internal.config.config import Config
+
+
+class ConfigAction(ABC):
+    """Common interface for all the actions."""
+
+    @abstractmethod
+    def apply(self, config: "Config") -> "Config":
+        """Interface for apply method."""
+        ...
+
+    @abstractclassmethod
+    def build(cls, key: str, data: Any) -> "ConfigAction":
+        """Interface for build method."""
+        ...
+
+    @staticmethod
+    def from_config_entry(key: str, data: Any) -> "ConfigAction":
+        """Config entries overwrite existing settings."""
+        return SetConfigValueAction.build(key, data)
+
+    @staticmethod
+    def from_target_override_entry(key: str, data: Any) -> "ConfigAction":
+        """Target overrides come in three shapes: addition, removal and override."""
+        # TODO: implement add/remove parsed out from the key
+        return SetConfigValueAction.build(key, data)
+
+
+@dataclass
+class SetConfigValueAction(ConfigAction):
```

Comment:
I think this name is a tad confusing. `Set` is an action so I would expect it to just be a function.  The class seems to be have the same purpose as `ConfigAction` but a level lower. Could it be `ConfigValueAction`?

Reply:
Without a few more things that inherit from `ConfigAction` I'm not sure what kind of think a subclass would be and how this subclass fits into it. Would you mind fleshing out the reasoning for it?

Reply:
I don't necessarily agree about naming, as there's two schools of thought - you can call your classes (services?) with _verbs_ or _nouns_. `ConfigValueSetter` vs `SetConfigValue`. Both can implement `apply/call` which would execute the action.

Your point about it being a function is interesting though, as python provides callable interface, which can potentially satisfy your naming concerns and simplify `apply` into `__call__`.

---
 
I wanted `ConfigAction` to be an interface/type that all actions adhere to. The requirement is:
- having a `build` classmethod which creates instances of self based on information parsed out from key and data
- having an `apply` that takes the config

It might be that it's not necessary. 🤔 

---

Further `ConfigAction`s would be:
- `AddConfigValue` (or `AppendConfigValue`) for things like `feature_add`
- `Remove*` - as above

Reply:
So a 'ConfigAction' is an action performed on the 'Config', not something the 'Config' contains. How would you feel about calling it 'ConfigModifier' or something like that instead of? Action is quite a vague term and it wasn't immediately clear what they were doing.

Reply:
https://github.com/ARMmbed/mbed-build/pull/18/commits/15dfd842d0425ac589ed5dd1dd3dc9171ea5dd60

Reply:
I like the modifier suggestion, I shall rename things.

Reply:
Going further, I think `ConfigLayer` is just another modifier (apply a set of modifiers), but I'm tempted to leave it as is for now. 😅

Reply:
Might be harder to undo further down the line. (Speaking as someone who has just had to make 1000+ lines of changes because of two changing names).

Reply:
This is very much internal to `mbed-build` so I doubt it'd have such 💥 effect. 👍

Reply:
Okey dokey 👍